sdk/src/main/java/com/mapbox/maps/renderer/FpsManager.kt

```diff
@@ -0,0 +1,186 @@
+
+  fun setScreenRefreshRate(screenRefreshRate: Int) {
+    if (this.screenRefreshRate == screenRefreshRate) {
+      return
+    }
+    this.screenRefreshRate = screenRefreshRate
+    screenRefreshPeriodNs = ONE_SECOND_NS / screenRefreshRate
+    if (userRefreshRate != USER_DEFINED_REFRESH_RATE_NOT_SET) {
+      userDefinedRatio = userRefreshRate.toDouble() / screenRefreshRate
+      logI(TAG, "User defined ratio is $userDefinedRatio")
+    }
+    if (LOG_STATISTICS) {
+      logI(
+        TAG,
+        "Device screen frequency is $screenRefreshRate," +
+          " user defined refresh rate is ${if (userRefreshRate == USER_DEFINED_REFRESH_RATE_NOT_SET) "not set" else userRefreshRate}"
+      )
+    }
+  }
+
+  fun setUserRefreshRate(refreshRate: Int) {
+    if (userRefreshRate != refreshRate) {
+      userRefreshRate = refreshRate
+      logI(TAG, "User set max FPS to $userRefreshRate")
+      if (screenRefreshRate != SCREEN_METRICS_NOT_DEFINED) {
+        userDefinedRatio = userRefreshRate.toDouble() / screenRefreshRate
+        logI(TAG, "User defined ratio is $userDefinedRatio")
+      }
+    }
+  }
+
+  /**
+   * Return true if rendering should happen this frame and false otherwise.
+   *
+   * @param frameTimeNs time value from Choreographer in [System.nanoTime] timebase.
+   */
+  fun preRender(frameTimeNs: Long): Boolean {
+    // clear any scheduled task as new render call is about to happen
+    handler.removeCallbacksAndMessages(fpsManagerToken)
+    // no need to perform neither pacing nor FPS calculation when setMaxFps / setOnFpsChangedListener was not called by the user
+    if (userDefinedRatio == USER_DEFINED_RATIO_NOT_SET && fpsChangedListener == null) {
+      return true
+    }
+    preRenderTimeNs = System.nanoTime()
+    var skippedNow = 0
+    // check if we did miss VSYNC deadline meaning too much work was done in previous doFrame
+    if (previousRefreshNs != -1L && frameTimeNs - previousRefreshNs > screenRefreshPeriodNs + ONE_MILLISECOND_NS) {
+      skippedNow =
+        ((frameTimeNs - previousRefreshNs) / (screenRefreshPeriodNs + ONE_MILLISECOND_NS)).toInt()
+      choreographerSkips += skippedNow
+    }
+    previousRefreshNs = frameTimeNs
+    // we always increase choreographer tick by one + add number of skipped frames for consistent results
+    choreographerTicks += skippedNow + 1
+    if (skippedNow > 0 && LOG_STATISTICS) {
+      logW(
+        TAG,
+        "Skipped $skippedNow VSYNC pulses since last actual render," +
+          " total skipped in measurement period $choreographerSkips / $choreographerTicks"
+      )
+    }
+    if (userDefinedRatio != USER_DEFINED_RATIO_NOT_SET) {
+      return performPacing()
+    }
+    return true
+  }
+
+  fun postRender() {
+    fpsChangedListener?.let { listener ->
+      val frameRenderTimeNs = System.nanoTime() - preRenderTimeNs
+      frameRenderTimeAccumulatedNs += frameRenderTimeNs
+      // normally we update FPS counter once a second
```

Comment:
Where is this `every second` checked?

Reply:
cause `every second` is actually when number of VSYNC impulses is equal to given screen frequency. So when we have 60 choreographer ticks - it means that 1 second did pass.